packages/parser/src/language/sankey/sankeyMatcher.ts

```diff
@@ -0,0 +1,14 @@
+/**
+ * Matches sankey link source node
+ */
+export const sankeyLinkSourceRegex = /(?:"((?:""|[^"])+)")|([^\n\r,]+(?=%%)|[^\n\r,]+)/;
+
+/**
+ * Matches sankey link target node
+ */
+export const sankeyLinkTargetRegex = /,(?:(?:"((?:""|[^"])+)")|([^\n\r,]+))/;
+
+/**
+ * Matches sankey link value
+ */
+export const sankeyLinkValueRegex = /,("(?:0|[1-9]\d*)(?:\.\d+)?"|[\t ]*(?:0|[1-9]\d*)(?:\.\d+)?)/;
```

Comment:
- That is not clear why commas should be part of these regexps, while they are very likely must be a part of grammar (not necessarily), but they are not necessary here
- In our first approach we allowed spaces around values, this was our extension of CSV because basically this syntax is CSV format apart from
  - additional spaces around values and at the beginning and at the end of strings
  - allowed empty lines

Reply:
> * That is not clear why commas should be part of these regexps, while they are very likely must be a part of grammar (not necessarily), but they are not necessary here

They are part of the grammar.

> Matchers (most if the time) are just regexes to extract wanted data from tokens.
>
> https://github.com/mermaid-js/mermaid/pull/4799#discussion_r1313973878

So for example, the lexer would tokenize the follow code:

```
wow, awesome, 10
```

into:

```
wow       ==> SANKEY_LINK_SOURCE
, awesome ==> SANKEY_LINK_TARGET
, 10      ==> SANKEY_LINK_VALUE
```

So I created these regexes that extract the data.
They could be replaced with some if-else and slicing.


>   * additional spaces around values


The rule already matches spaces before values, and there a hidden rule that ignores space at the end of values

https://github.com/mermaid-js/mermaid/blob/b6906a6474e01b5a6946749700150a4bada61657/packages/parser/src/language/sankey/sankey.langium#L20

> and at the beginning and at the end of strings

If I understand correctly, the rule would match them. Could you give an example?

Reply:
we have to ensure that
```
 a , b , 10 
 "a" , "b" , 123
```
works

Yes, I see that this is part of lexer.
Initial grammar was like:
```
node comma node comma value
```
now grammar becomes
```
node node value
```
I am not against this, but it seems to me that regexp grabs something that is not related directly to the node, or value (comma). Comma is simply a delimiter, so the first approach seems more correct to me.

Reply:
> we have to ensure that
> 
> ```
>  a , b , 10 
>  "a" , "b" , 123
> ```
> 
> works

In CSV, the second link would be:

| first column | second column | third column |
|:-:|:-:|:-:|
| `"a"` | `"b"` | `123` |

Not just strings without doable quotations, and it's the current approach.

I'm not sure what you mean by insure, we already have test cases that cover that.

> Yes, I see that this is part of lexer. Initial grammar was like:
> 
> ```
> node comma node comma value
> ```
> 
> now grammar becomes
> 
> ```
> node node value
> ```
> 
> I am not against this, but it seems to me that regexp grabs something that is not related directly to the node, or value (comma). Comma is simply a delimiter, so the first approach seems more correct to me.
> 
> From my perspective it would be better to create a matcher for a comma

Originally, we had a rule for comma, but I removed it.

In the jison parser, it uses states to shorten possible rules within the current context, e.g., if a `source` has been matched, then the possible rule after it is a comma; if it doesn't find a comma, it throws an error.
In langium parser (or chevrotain to specific), it has modes, it's similar states in jison, but I wouldn't recommend using it unless we actually needs to.

Anyway, if we created a rule for commas, how would we be able to distinguish between a `source` and a `target` rules? Both of them would have identical rules; they both match everything except `,` and newlines. We can't use context without modes in langium.

Reply:
In case you may wonder why I'm against using modes, it's because the default parser or lexer errors might be weird and not understandable.
And modes aren't part of grammar (`.langium` file); they're implemented in `TokenBuilder`, which would diverge and make it more complicated to adjust more services for straight-forward grammar.

Reply:
In jison we can differ them by giving different names, or [auto-numbers](https://gerhobbelt.github.io/jison/docs/#auto-numbered-named-accessors). So the same token can appear more than once. I am not familiar with the languim, did not have time to check it. So I rely on your decision in that case. I think we will reconsider and review everything one more time when releasing new version

Reply:
I think better example is a classic one:
```
expression ::= number '+' number | number '-' number;
```
left and the right operands are recognized by the same regular expression and there is not difference between them.
`+` and `-` are not the part of the numbers, unless there is additional unary operator.

Another example is:
```
assignment :== variable '=' variable | variable '=' expression;
```
There is no difference in terms of lexems. Both, left part of assignment and right part of assignment can be the same lexem. '=' is not the sub-element of the right part of the assignment. Parsers usually add additional attributes to lexems  like exact match, value etc.

So my point is that when you recognize `source` and `target` you are still recognizing `node`, they are identical, and their regular expressions are identical.
Otherwise we could as well use one regexp and capture groups  to recognize the whole string.

Reply:
This is really strange that we cannot differ them. A rule like
```
line ::= number number number
```
is absolutely valid. And we must be able to get 3 values out of it. If languim does not allow this, may be it is not the way to go.